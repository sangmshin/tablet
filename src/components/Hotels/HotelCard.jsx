import React, { memo, createRef, useState, useEffect } from 'react';
import { Image, Row, Col } from 'react-bootstrap';
import style from './Hotels.module.scss';
import PropTypes from "prop-types";
import Caption from './Caption';
import heart from './heart-selected.svg';
import unheart from './heart-unselected.svg';

const HotelCard = memo( ({ source, bookmark, unbookmark }) => {
  // console.log(source);
  const [ isAnimation, setAnimation ] = useState(true)
  const [ isVisible, setVisible ] = useState(false)
  const caption = createRef();
  const hotel_img = createRef();

  const onResize = e => {
    window.innerWidth < 992 && window.innerWidth > 575
    ? setAnimation(false)
    : setAnimation(true)
  }

  useEffect(()=> {

    window.innerWidth < 992 && window.innerWidth > 575 
    && setAnimation(false)

    window.addEventListener('resize', onResize)

    return () => window.removeEventListener('resize', onResize)

  }, [])

  const animate = direction => {
    if (!isAnimation) return;

    const _img = hotel_img.current
    const _caption = caption.current.firstElementChild;

    if (direction === 'up') {
      _img.classList.add(style.zeroHeight)
      _caption.classList.add(style.fullHeight)
      setVisible(true)
    } else {
      _img.classList.remove(style.zeroHeight)
      _caption.classList.remove(style.fullHeight)
      setVisible(false)
    }
  }

  const captionLists = [
    {
      type: source.criteria.atmosphere, 
      visibility: true
    },
    {
      type: source.criteria.style, 
      visibility: true
    },
    {
      type: source.address[0], 
      visibility: isVisible
    },
    {
      type: source.amenities[0], 
      visibility: isVisible
    }, 
    {
      type: source.phone, 
      visibility: isVisible
    }
  ]
  
  return (
    <Col xs={12} sm={6} md={4} lg={4} xl={4} 
      onMouseOver={()=> animate('up')} 
      onMouseOut={()=> animate('down')} 
      className={style.card}
    >
      <div className={style.card_wrapper} ref={hotel_img}>
        <Image 
          className={style.hotelImg} 
          src={source.images && `https://${source.images[0].hotrooms_large_url}`} 
          alt="Hotel Image" fluid 
        />

        <div className={style.hotel_name}>
          <h1>{source.hotel_name}</h1>
        </div>

        <Image 
          onClick={source.bookmarked ? unbookmark : bookmark} 
          className={style.heart} 
          src={source.bookmarked ? heart : unheart} 
          width={"40px"}
        />

      </div>
      <div ref={caption}>
        <Row className={style.detail_wrapper}>
          {
            captionLists.map( ({type, visibility}, i) => 
              <Caption type={type} visible={visibility} key={i}/>
            )
          }
        </Row>
      </div>
    </Col>
  )
})


HotelCard.propTypes = {
  source: PropTypes.object.isRequired,
  bookmark: PropTypes.func.isRequired,
  unbookmark: PropTypes.func.isRequired,
};

export default HotelCard;
