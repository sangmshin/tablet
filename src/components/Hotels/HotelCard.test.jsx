import React from 'react';
import HotelCard from './HotelCard';
import { shallow, render, mount } from 'enzyme';

it("renders correctly", () => {
  const wrapper = shallow(
    // eslint-disable-next-line react/react-in-jsx-scope
    <HotelCard source={{address:['bergen rd.'], amenities:['gym'], phone:"9173276941", criteria:{atmosphere:"lively", style:"modern"}}} bookmark={()=>{}} unbookmark={()=>{}}/>
  )
  // expect(wrapper).toMatchSnapshot()
})

it("renders correctly 2", () => {
  const wrapper = render(
    // eslint-disable-next-line react/react-in-jsx-scope
    <HotelCard source={{address:['bergen rd.'], amenities:['gym'], phone:"9173276941", criteria:{atmosphere:"lively", style:"modern"}}} bookmark={()=>{}} unbookmark={()=>{}}/>
  )
  // expect(wrapper).toMatchSnapshot()
})

it("renders correctly 3", () => {
  const wrapper = mount(
    // eslint-disable-next-line react/react-in-jsx-scope
    <HotelCard source={{address:['bergen rd.'], amenities:['gym'], phone:"9173276941", criteria:{atmosphere:"lively", style:"modern"}}} bookmark={()=>{}} unbookmark={()=>{}}/>
  )

  const text = wrapper.find("Caption").at(1).find("p").text()
  console.log(text);
  
  // expect(wrapper).toMatchSnapshot()
})