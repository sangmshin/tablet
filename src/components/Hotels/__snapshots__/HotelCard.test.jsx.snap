// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Col
  as="div"
  className="card"
  lg={4}
  md={4}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  sm={6}
  xl={4}
  xs={12}
>
  <div
    className="card_wrapper"
  >
    <Bootstrap(Image)
      alt="Hotel Image"
      className="hotelImg"
      fluid={true}
    />
    <div
      className="hotel_name"
    >
      <h1 />
    </div>
    <Bootstrap(Image)
      className="heart"
      src="heart-unselected.svg"
      width="40px"
    />
  </div>
  <div>
    <Bootstrap(Row)
      className="detail_wrapper"
    >
      <Caption
        type="lively"
        visible={true}
      />
      <Caption
        type="modern"
        visible={true}
      />
      <Caption
        key="0"
        type="bergen rd."
        visible={false}
      />
      <Caption
        key="1"
        type="gym"
        visible={false}
      />
      <Caption
        key="2"
        type="9173276941"
        visible={false}
      />
    </Bootstrap(Row)>
  </div>
</Col>
`;

exports[`renders correctly 2 1`] = `
<div
  class="card col-xl-4 col-lg-4 col-md-4 col-sm-6 col-12"
>
  <div
    class="card_wrapper"
  >
    <img
      alt="Hotel Image"
      class="hotelImg img-fluid"
    />
    <div
      class="hotel_name"
    >
      <h1 />
    </div>
    <img
      class="heart"
      src="heart-unselected.svg"
      width="40px"
    />
  </div>
  <div>
    <div
      class="detail_wrapper row"
    >
      <div
        class="col-lg-6 col-md-12 col-sm-12 col-6"
      >
        <p
          class="criteria"
        >
          lively
        </p>
      </div>
      <div
        class="col-lg-6 col-md-12 col-sm-12 col-6"
      >
        <p
          class="criteria"
        >
          modern
        </p>
      </div>
      <div
        class="col-lg-6 col-md-12 col-sm-12 col-6"
      />
      <div
        class="col-lg-6 col-md-12 col-sm-12 col-6"
      />
      <div
        class="col-lg-6 col-md-12 col-sm-12 col-6"
      />
    </div>
  </div>
</div>
`;
